### ISSUE TITLE
Problem with hash_behaviour=merge and dynamic groups

### ISSUE THREAD

**Issue by geraldh** [ansible/ansible#9085]
##### Issue Type:

Bug Report
##### Ansible Version:

1.6.10 and 1.7.1
##### Environment:
- Ubuntu 14.04
- python-2.7.6
- ansible.cfg: hash_behaviour=merge
##### Summary:

When merging a dict defined in a dynamic group_var with a dict defined in a static host_var / group_var the content of the dict of the dynamic group_var is missing in the resulting merged dict in 1.7.1

In 1.6.10 the merged dict correctly contains the values from both dynamic and static group/host_vars.
##### Steps To Reproduce:

**hosts**

```
[linuxServers]
linux1.example.com
```

**group_vars/ansible-os-family-RedHat**

```

---
authorized_keys:
  root:
    user1:
    user2:
```

**host_vars/linux1.example.com**

```

---
authorized_keys:
  root:
    user3:
```

**site.yml**

```

---
- hosts: linux1.example.com
  tasks:

    - name: group_by ansible_os_family
      group_by: key=ansible-os-family-{{ ansible_os_family }}
      changed_when: false

    - name: output dictionary
      debug: var=authorized_keys
```

The intention is to sum up all authorized users in different group_vars / host_vars with hash_behavior=merge and finally let a template create the authorized_keys file.

The host `linux1.example.com` is a RedHat host so both the dynamic group_var `ansible-os-family-RedHat` and the static host_vars  from `linux1.example.com` apply.
##### Expected Results:

Running the above playbook on **1.6.10**

```
TASK: [output dictionary] ********
ok: [linux1.example.com] => {
    "authorized_keys": {
        "root": {
            "user1": null, 
            "user2": null, 
            "user3": null
        }
    }
}
```
##### Actual Results:

Running on **1.7.1**

```
TASK: [output dictionary] ******** 
ok: [linux1.example.com] => {
    "authorized_keys": {
        "root": {
            "user3": null
        }
    }
}
```
- The dictionary of the dynamic group `ansible-os-family-RedHat` is _missing_ on 1.7.1 !
- This does not happen if you use a static group_var instead of a dynamic one.

I think this is not the expected behaviour and this could have catastrophic consequences if you intend to use the dict merge for generating authorized_keys - you could lock out yourself from key based ssh access to all ansible manged hosts with one playbook run after upgrading ansible from 1.6.10 to 1.7.1

Please investigate.

Thanks,
Gerald

**Comment by srgvg** [comment]
I can confirm this bug, but can't reproduce it wokring as expected in 1.6.10.
I gisted my test files here, same as @geraldh except working on local Ubuntu

**Comment by srgvg** [comment]
Sorry, forgot to confirm hash_behaviour merge, added an ansible.cfg to the gists. I can confirm the issue, but also, it is already fixed in 1.8 devel
@geraldh can you confirm?

**Comment by geraldh** [comment]
@sergevanginderachter - yes works as expected in 1.8 devel for me also !

**Comment by srgvg** [comment]
(FYI @jimi-c the fix was in a1f09bd89f6daec9365f0654bda6e4e9654be45c - might be a candidate for 1.7.2)

**Comment by jimi-c** [comment]
@sergevanginderachter thanks, I've noted the commit so if we do release a 1.7.2 I can include this.

If you continue seeing any problems related to this issue, or if you have any further questions, please let us know by stopping by one of the two mailing lists, as appropriate:
- https://groups.google.com/forum/#!forum/ansible-project - for user questions, tips, and tricks
- https://groups.google.com/forum/#!forum/ansible-devel - for strategy, future planning, and questions about writing code

Because this project is very active, we're unlikely to see comments made on closed tickets, but the mailing list is a great way to ask questions, or post if you don't think this particular issue is resolved.

Thank you!
